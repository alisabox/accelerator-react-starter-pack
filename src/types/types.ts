import { Action } from 'redux';
import { ThunkAction}  from 'redux-thunk';
import { AxiosInstance } from 'axios';
import { RootState } from '../store/reducers/root-reducer';
import { SeachOptions } from '../const/const';

export type GuitarType = {
  id: number;
  name: string;
  vendorCode: string;
  type: string;
  description: string;
  previewImg: string;
  stringCount: number;
  rating: number;
  price: number;
}

export type CommentType = {
  id: string;
  userName: string;
  advantages: string;
  disadvantages: string;
  comment: string;
  rating: number;
  createAt: Date;
  guitarId: number;
}

export type FilterSettingsType = {
  minPrice: number;
  maxPrice: number;
}

export type GuitarAndCommentsType = GuitarType & {
  comments: CommentType[],
}

export type guitarsStateType = {
  guitarsAndComments: GuitarAndCommentsType[],
  guitarsPerPage: GuitarAndCommentsType[],
  searchResult: GuitarType[],
}

export type filterStateType = {
  filterURLOptions: SeachOptionsType,
  minPrice?: number,
  maxPrice?: number,
}

export type SeachOperatorsBiasedType = {
  [key: string]: string;
}

export type SortOperatorsType = {
  [key: string]: string;
}


export type FilterParamsType = {
  filters: string,
  sort: string,
  order: string,
  pageSearch: string,
  page: string,
}

export type SeachOptionsType = {
  [SeachOptions.PriceMin]: string,
  [SeachOptions.PriceMax]: string,
  [SeachOptions.Acoustic]: boolean,
  [SeachOptions.Electric]: boolean,
  [SeachOptions.Ukulele]: boolean,
  [SeachOptions.FourStrings]: boolean,
  [SeachOptions.SixStrings]: boolean,
  [SeachOptions.SevenStrings]: boolean,
  [SeachOptions.TwelveStrings]: boolean,
}

export type State = RootState;
export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, State, AxiosInstance, Action>;
export type ReselectType<S> = (state: State) => S;
