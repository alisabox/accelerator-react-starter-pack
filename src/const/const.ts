import { SeachOperatorsBiasedType, SortOperatorsType } from '../types/types';

export const APIRoute = {
  Guitars: '/guitars',
  Comments: '/comments',
  Coupons: '/coupons',
  Orders: '/orders',
} as const;

export const AppRoute = {
  Root: '/',
} as const;

export enum ActionType {
  GetGuitarsAndComments = 'MAIN/GET_GUITARS_AND_COMMENTS',
  GetGuitarsPerPage = 'MAIN/GET_GUITARS_PER_PAGE',
  GetSearchResult = 'MAIN/GET_SEARCH_RESULT',
  ClearSearchResult = 'MAIN/CLEAR_SEARCH_RESULT',
  GetFilterResult = 'MAIN/GET_FILTER_RESULT',
  GetFilterURLOptions = 'MAIN/GET_FILTER_URL_OPTIONS',
  GetMinPrice = 'MAIN/GET_MIN_PRICE',
  GetMaxPrice = 'MAIN/GET_MAX_PRICE',
}

export enum CatalogSort {
  Price = 'price',
  Rating = 'rating',
  None = 'none',
}

export enum CatalogSortOrder {
  Down = 'desc',
  Up = 'asc',
  None = 'none',
}

export const CatalogSortOperators: SortOperatorsType = {
  [CatalogSort.Price]: '_sort=price',
  [CatalogSort.Rating]: '_sort=rating',
  [CatalogSortOrder.Down]: '_order=desc',
  [CatalogSortOrder.Up]: '_order=asc',
} as const;

export const SeachOperators = {
  MinPrice: 'price_gte=',
  MaxPrice: 'price_lte=',
  TypeAcoustic: 'type=acoustic',
  TypeElectric: 'type=electric',
  TypeUkulele: 'type=ukulele',
  FourStrings: 'stringCount=4',
  SixStrings: 'stringCount=6',
  SevenStrings: 'stringCount=7',
  TwelveStrings: 'stringCount=12',
} as const;

export enum SeachOptions {
  PriceMin = 'priceMin',
  PriceMax = 'priceMax',
  Acoustic = 'acoustic',
  Electric = 'electric',
  Ukulele = 'ukulele',
  FourStrings = '4-strings',
  SixStrings = '6-strings',
  SevenStrings = '7-strings',
  TwelveStrings = '12-strings',
}

export const SeachOperatorsBiased: SeachOperatorsBiasedType = {
  [SeachOptions.PriceMin]: SeachOperators.MinPrice,
  [SeachOptions.PriceMax]: SeachOperators.MaxPrice,
  [SeachOptions.Acoustic]: SeachOperators.TypeAcoustic,
  [SeachOptions.Electric]: SeachOperators.TypeElectric,
  [SeachOptions.Ukulele]: SeachOperators.TypeUkulele,
  [SeachOptions.FourStrings]: SeachOperators.FourStrings,
  [SeachOptions.SixStrings]: SeachOperators.SixStrings,
  [SeachOptions.SevenStrings]: SeachOperators.SevenStrings,
  [SeachOptions.TwelveStrings]: SeachOperators.TwelveStrings,
} as const;

export const NUMBER_OF_CARDS = 9;

export const ERROR_MESSAGE = 'Не удалось загрузить данные каталога. Попробуйте еще раз позже';

export enum EmptyCatalogMessage {
  Loading = 'Загружаем...',
  ConnectionError = 'Не удалось загрузить каталог. Проверьте подключение к интернету.',
}
