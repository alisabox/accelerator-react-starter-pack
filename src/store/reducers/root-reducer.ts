import { combineReducers } from 'redux';
import { filterReducer } from './filter/filter-reducer';
import { guitarsReducer } from './guitars/guitars-reducer';

export enum NameSpace {
  Guitars = 'Guitars',
  Filter = 'Filter',
}

export const rootReducer = combineReducers({
  [NameSpace.Guitars]: guitarsReducer,
  [NameSpace.Filter]: filterReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
