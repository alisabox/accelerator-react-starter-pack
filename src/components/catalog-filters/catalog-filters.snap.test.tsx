import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { configureMockStore } from '@jedmao/redux-mock-store';
import { createMemoryHistory } from 'history';
import { createAPI } from '../../services/api';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { State } from '../../types/types';
import CatalogFilters from './catalog-filters';
import { guitars } from '../../mocks/mocks';
import { NameSpace } from '../../store/reducers/root-reducer';
import { SeachOptions } from '../../const/const';

const history = createMemoryHistory();

const mockData = guitars;

const api = createAPI();
const middlewares = [thunk.withExtraArgument(api)];
const mockStore = configureMockStore<
    State,
    Action,
    ThunkDispatch<State, typeof api, Action>
  >(middlewares);

const store = mockStore({
  [NameSpace.Guitars]: {
    guitarsAndComments: mockData,
    guitarsPerPage: [],
  },
  [NameSpace.Filter]: {
    filterURLOptions: {
      [SeachOptions.PriceMin]: '',
      [SeachOptions.PriceMax]: '',
      [SeachOptions.Acoustic]: false,
      [SeachOptions.Electric]: false,
      [SeachOptions.Ukulele]: false,
      [SeachOptions.FourStrings]: false,
      [SeachOptions.SixStrings]: false,
      [SeachOptions.SevenStrings]: false,
      [SeachOptions.TwelveStrings]: false,
    },
  },
});
describe('Component: Catalog filters', () => {
  test('should render correctly vs. snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Router history={history}>
          <CatalogFilters />
        </Router>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
