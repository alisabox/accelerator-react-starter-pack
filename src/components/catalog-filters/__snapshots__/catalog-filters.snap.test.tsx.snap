// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Catalog filters should render correctly vs. snapshot 1`] = `
<div>
  <form
    class="catalog-filter"
  >
    <h2
      class="title title--bigger catalog-filter__title"
    >
      Фильтр
    </h2>
    <fieldset
      class="catalog-filter__block"
    >
      <legend
        class="catalog-filter__block-title"
      >
        Цена, ₽
      </legend>
      <div
        class="catalog-filter__price-range"
      >
        <div
          class="form-input"
        >
          <label
            class="visually-hidden"
          >
            Минимальная цена
          </label>
          <input
            id="priceMin"
            name="priceMin"
            type="number"
            value=""
          />
        </div>
        <div
          class="form-input"
        >
          <label
            class="visually-hidden"
          >
            Максимальная цена
          </label>
          <input
            id="priceMax"
            name="priceMax"
            type="number"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <fieldset
      class="catalog-filter__block"
    >
      <legend
        class="catalog-filter__block-title"
      >
        Тип гитар
      </legend>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="acoustic"
          name="acoustic"
          type="checkbox"
        />
        <label
          for="acoustic"
        >
          Акустические гитары
        </label>
      </div>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="electric"
          name="electric"
          type="checkbox"
        />
        <label
          for="electric"
        >
          Электрогитары
        </label>
      </div>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="ukulele"
          name="ukulele"
          type="checkbox"
        />
        <label
          for="ukulele"
        >
          Укулеле
        </label>
      </div>
    </fieldset>
    <fieldset
      class="catalog-filter__block"
    >
      <legend
        class="catalog-filter__block-title"
      >
        Количество струн
      </legend>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="4-strings"
          name="4-strings"
          type="checkbox"
        />
        <label
          for="4-strings"
        >
          4
        </label>
      </div>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="6-strings"
          name="6-strings"
          type="checkbox"
        />
        <label
          for="6-strings"
        >
          6
        </label>
      </div>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="7-strings"
          name="7-strings"
          type="checkbox"
        />
        <label
          for="7-strings"
        >
          7
        </label>
      </div>
      <div
        class="form-checkbox catalog-filter__block-item"
      >
        <input
          class="visually-hidden"
          id="12-strings"
          name="12-strings"
          type="checkbox"
        />
        <label
          for="12-strings"
        >
          12
        </label>
      </div>
    </fieldset>
  </form>
</div>
`;
